Mathlib/CategoryTheory/Limits/IsLimit.lean

```diff
@@ -0,0 +1,1085 @@
+/-
+Copyright (c) 2018 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Reid Barton, Mario Carneiro, Scott Morrison, Floris van Doorn
+
+! This file was ported from Lean 3 source module category_theory.limits.is_limit
+! leanprover-community/mathlib commit 740acc0e6f9adf4423f92a485d0456fc271482da
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Adjunction.Basic
+import Mathlib.CategoryTheory.Limits.Cones
+
+/-!
+# Limits and colimits
+
+We set up the general theory of limits and colimits in a category.
+In this introduction we only describe the setup for limits;
+it is repeated, with slightly different names, for colimits.
+
+The main structures defined in this file is
+* `IsLimit c`, for `c : Cone F`, `F : J ⥤ C`, expressing that `c` is a limit cone,
+
+See also `CategoryTheory.Limits.HasLimits` which further builds:
+* `LimitCone F`, which consists of a choice of cone for `F` and the fact it is a limit cone, and
+* `HasLimit F`, asserting the mere existence of some limit cone for `F`.
+
+## Implementation
+At present we simply say everything twice, in order to handle both limits and colimits.
+It would be highly desirable to have some automation support,
+e.g. a `@[dualize]` attribute that behaves similarly to `@[to_additive]`.
+
+## References
+* [Stacks: Limits and colimits](https://stacks.math.columbia.edu/tag/002D)
+
+-/
+
+
+noncomputable section
+
+open CategoryTheory CategoryTheory.Category CategoryTheory.Functor Opposite
+
+namespace CategoryTheory.Limits
+
+-- declare the `v`'s first; see `CategoryTheory.Category` for an explanation
+universe v₁ v₂ v₃ v₄ u₁ u₂ u₃ u₄
+
+variable {J : Type u₁} [Category.{v₁} J] {K : Type u₂} [Category.{v₂} K]
+
+variable {C : Type u₃} [Category.{v₃} C]
+
+variable {F : J ⥤ C}
+
+/-- A cone `t` on `F` is a limit cone if each cone on `F` admits a unique
+cone morphism to `t`.
+
+See <https://stacks.math.columbia.edu/tag/002E>.
+  -/
+-- Porting note: removed @[nolint has_nonempty_instance]
+structure IsLimit (t : Cone F) where
+  /-- There is a morphism from any cone vertex to `t.pt` -/
+  lift : ∀ s : Cone F, s.pt ⟶ t.pt
+  /-- The map makes the triangle with the two natural transformations commute -/
+  fac : ∀ (s : Cone F) (j : J), lift s ≫ t.π.app j = s.π.app j := by aesop_cat
+  /-- It is the unique such map to do this -/
+  uniq : ∀ (s : Cone F) (m : s.pt ⟶ t.pt) (_ : ∀ j : J, m ≫ t.π.app j = s.π.app j), m = lift s := by
+    aesop_cat
+#align category_theory.limits.is_limit CategoryTheory.Limits.IsLimit
+#align category_theory.limits.is_limit.fac' CategoryTheory.Limits.IsLimit.fac
+#align category_theory.limits.is_limit.uniq' CategoryTheory.Limits.IsLimit.uniq
+
+-- Porting note: linter claimed it reduced but it did not
+attribute [nolint simpNF] IsLimit.mk.injEq
```

Comment:
It's silly, yes, but the LHS of this lemma does reduce (to `True`) due to the subsingleton instance on line 79. 
I don't know what the right fix is here.

(Similarly for the colimit case.)

Reply:
`simp` can prove this